app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt

```diff
@@ -0,0 +1,470 @@
+     * If data is not cached, tiles containing the [bbox] are fetched from database and cached.
+     */
+    fun getMapDataWithGeometry(bbox: BoundingBox): MutableMapDataWithGeometry = synchronized(this) {
+        val requiredTiles = bbox.enclosingTilesRect(tileZoom).asTilePosSequence().toList()
+        val cachedTiles = spatialCache.getTiles()
+        val tilesToFetch = requiredTiles.filterNot { it in cachedTiles }
+        val tilesRectToFetch = tilesToFetch.minTileRect()
+
+        val result = MutableMapDataWithGeometry()
+        result.boundingBox = bbox
+        val nodes: Collection<Node>
+        if (tilesRectToFetch != null) {
+            // fetch needed data
+            val fetchBBox = tilesRectToFetch.asBoundingBox(tileZoom)
+            val (elements, geometries) = fetchMapData(fetchBBox)
+
+            // get nodes from spatial cache
+            // this may not contain all nodes, but tiles that were cached initially might
+            // get dropped when the caches are updated
+            // duplicate fetch might be unnecessary in many cases, but it's very fast anyway
+
+            nodes = HashSet<Node>(spatialCache.get(bbox))
+            update(updatedElements = elements, updatedGeometries = geometries, bbox = fetchBBox)
+
+            // return data if we need exactly what was just fetched
+            if (fetchBBox == bbox) {
+                val nodeGeometryEntries = elements.filterIsInstance<Node>().map { it.toElementGeometryEntry() }
+                result.putAll(elements, geometries + nodeGeometryEntries)
+                return result
+            }
+
+            // get nodes again, this contains the newly added nodes, but maybe not the old ones if cache was trimmed
+            nodes.addAll(spatialCache.get(bbox))
+        } else {
+            nodes = spatialCache.get(bbox)
+        }
+
+        val wayIds = HashSet<Long>(nodes.size / 5)
+        val relationIds = HashSet<Long>(nodes.size / 10)
+        for (node in nodes) {
+            wayIdsByNodeIdCache[node.id]?.let { wayIds.addAll(it) }
+            relationIdsByElementKeyCache[ElementKey(ElementType.NODE, node.id)]?.let { relationIds.addAll(it) }
+            result.put(node, ElementPointGeometry(node.position))
+        }
+        for (wayId in wayIds) {
+            result.put(wayCache[wayId]!!, wayGeometryCache[wayId])
+            relationIdsByElementKeyCache[ElementKey(ElementType.WAY, wayId)]?.let { relationIds.addAll(it) }
+        }
+        for (relationId in relationIds) {
+            result.put(relationCache[relationId]!!, relationGeometryCache[relationId])
+            // don't add relations of relations, because elementDao.getAll(bbox) also isn't doing that
+        }
+
+        // trim if we fetched new data, and spatialCache is full
+        // trim to 90%, so trim is (probably) not immediately called on next fetch
+        if (spatialCache.size >= maxTiles && tilesToFetch.isNotEmpty())
+            trim((maxTiles * 9) / 10)
```

Comment:
didn't you mention that this is expensive? If it is, I'd rather reduce the cache somewhat on one trim, so that the next trim is very long along the line.

This means that on average, the cache will be only half-full or something, but then you could just increase the total cache size to balance that out.

Reply:
Well, it's already trimmed to 90% if the cache is full, so it's unlikely that the next fetch will actually trim.
And I noticed that there is a _bug_, as `spatialCache.size` will include empty tiles, which are not trimmed. So with some empty tiles this will always return true if something was fetched from database.

The performance is still not great, but massively better than initially because
* No need for `spatialCache.getKeys()` when checking whether a node is cached. `spatialCache.get(id) != null` does not slow things down measurable, and doesn't require the slow `spatialCache.getKeys().toSet()`.
Looks like the `synchronized` on every call to `spatialCache.get(id)` doesn't slow down anything here...
* `getWayAndRelationIdsWithElementsInSpatialCache` is much faster than what I had used initially

So performance impact of `trimNonSpatialCaches` is mostly negligible compared to fetching from database, and trim is never called if all data is cached.
But still, we may actually want to keep elements cached that don't have any nodes/members in spatialCache.
So increasing the cache size, and trimming to 50% if size gets too large indeed looks like a good idea.

Reply:
Or how about checking the number of cached elements?
i.e. trim only if
* spatialCache is at maximum size
* data has just been fetched from db
* `wayCache.size` + `relationCache.size` + `spatialCache.byKey.size` > some threshold

Reply:
> And I noticed that there is a bug, as spatialCache.size will include empty tiles, which are not trimmed.

I did wonder about this. This exclusion of empty tiles for the whole size/trimming thing introduces some awkward complexity and I wonder if it is worth it. 
Maybe add at least an explanatory comment in the code why this is done and if you keep it, I suggest that `size` should return the non-empty tiles to properly encapsulate this behavior. (What is an empty tile anyway? How likely or relevant is it that there is really no data in a tile?)

> So increasing the cache size, and trimming to 50% if size gets too large indeed looks like a good idea.

I was thinking of 2/3 just because it's 1/3 for `trim()`

> Or how about checking the number of cached elements?
> i.e. trim only if
>    spatialCache is at maximum size
>    data has just been fetched from db
>    wayCache.size + relationCache.size + spatialCache.byKey.size > some threshold

Uhm I don't know, sounds a bit too complex to me for an unclear gain.

Reply:
> Maybe add at least an explanatory comment in the code why this is done and if you keep it, I suggest that `size` should return the non-empty tiles to properly encapsulate this behavior.

Will do. This also simplifies `SpatialCache.trim()`.

> What is an empty tile anyway?

A tile where no data is in the database (no nodes, or no quests).

> How likely or relevant is it that there is really no data in a tile?

Depends heavily on how SC is used. If you just have it centered at your own position you might never have empty tiles. But if you scroll around to decide where to go, or pre-load data you might have a lot of empty tiles.

Caching them has similar use as caching `null` elements/geometries.

> I was thinking of 2/3 just because it's 1/3 for trim()

Ok, so I'll increase the size by 50% and trim to 2/3 of the new max size.

Reply:
done in https://github.com/streetcomplete/StreetComplete/pull/4125/commits/63515ba363a2a2f6811a2e4d9437f773d2a6d3ec and https://github.com/streetcomplete/StreetComplete/pull/4125/commits/f83846d70e8a4adb240fb0bc2020034c509d2754